{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"padding-bottom:30px\">\n",
    "<a href=\"https://github.com/iqtk/iqtk\"><img src=\"https://raw.githubusercontent.com/iqtk/iqtk/master/inquiry/docs/assets/logotype_blue_small.png\" style=\"width:100px; margin-left:0px\"></img></a>\n",
    "<p style=\"color:#9E9E9E\">\n",
    "<a href=\"https://github.com/cwbeitel/inquiry/tree/master/docs\">Getting Started Guide</a> // <a href=\"https://goo.gl/forms/2cOmuUrQ3n3CKpim1\">Documentation Feedback</a></p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 style=\"color:#9E9E9E\">Genotype analysis with GATK</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this analysis we use the Broad Institute GATK toolset to make genotype calls for a sample of interest. Read more about [GATK](https://software.broadinstitute.org/gatk/) or [genotyping in general](https://en.wikipedia.org/wiki/Genotyping).\n",
    "\n",
    "To review, here we'll be running workflows generated by the Inquiry Toolkit directly from the Cloud DataFlow UI so once you've completed the [getting started documentation]() you'll be ready for this tutorial!\n",
    "\n",
    "Our data processing pipeline will deliver data to a BigQuery data table and we'll both explore our data from the BigQuery UI as well as pull a subset into this notebook for subsequent analysis and visualization."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 style=\"color:#9E9E9E\">Configuration and Run</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing we need to do is parameterize our analysis. The parameters file we'll need to provide should have the following structure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```json\n",
    "\n",
    "{\n",
    "  \"ref_fasta\": \"gs://cflow-public/data/genomes/Drosophila_melanogaster/Ensembl/BDGP5.25/Sequence/BWAIndex/genome.fa\",\n",
    "  \"reads\": [\n",
    "    [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794486_C2_R1_1_small.fq\",\n",
    "     \"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794486_C2_R1_2_small.fq\"],\n",
    "    [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794487_C2_R2_1_small.fq\",\n",
    "     \"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794487_C2_R2_2_small.fq\"],\n",
    "    [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794488_C2_R3_1_small.fq\",\n",
    "     \"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794488_C2_R3_2_small.fq\"]\n",
    "   ]\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Dataflow submission UI screenshot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once our job is submitted we'll be able to check its status using the Cloud DataFow UI as described in the [getting started documentation](). For illustration this workflow should look like the following:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Dataflow job running screenshot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When the workflow is complete you can obtain the path in Google Cloud Storage to the resulting files as described in the [getting started documentation]()."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 style=\"color:#9E9E9E\">Exploring the data</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we'll obtain some data resulting from the above workflow and play around a bit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>refname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>refbases</th>\n",
       "      <th>altbases</th>\n",
       "      <th>quality</th>\n",
       "      <th>filter</th>\n",
       "      <th>info</th>\n",
       "      <th>format</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>.</td>\n",
       "      <td>2L</td>\n",
       "      <td>18327</td>\n",
       "      <td>None</td>\n",
       "      <td>T</td>\n",
       "      <td>[&lt;, *, &gt;]</td>\n",
       "      <td>None</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=17;I16=15,2,0,0,680,27200,0,0,423,11439,0,0...</td>\n",
       "      <td>PL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>.</td>\n",
       "      <td>2L</td>\n",
       "      <td>10936</td>\n",
       "      <td>None</td>\n",
       "      <td>T</td>\n",
       "      <td>[&lt;, *, &gt;]</td>\n",
       "      <td>None</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=32;I16=20,12,0,0,1276,50896,0,0,1920,115200...</td>\n",
       "      <td>PL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.</td>\n",
       "      <td>2L</td>\n",
       "      <td>9858</td>\n",
       "      <td>None</td>\n",
       "      <td>T</td>\n",
       "      <td>[&lt;, *, &gt;]</td>\n",
       "      <td>None</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=30;I16=27,0,0,0,1200,57600,0,0,1620,97200,0...</td>\n",
       "      <td>PL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id refname  start   end refbases   altbases quality filter  \\\n",
       "0  .      2L  18327  None        T  [<, *, >]    None      .   \n",
       "1  .      2L  10936  None        T  [<, *, >]    None      .   \n",
       "2  .      2L   9858  None        T  [<, *, >]    None      .   \n",
       "\n",
       "                                                info format  \n",
       "0  DP=17;I16=15,2,0,0,680,27200,0,0,423,11439,0,0...     PL  \n",
       "1  DP=32;I16=20,12,0,0,1276,50896,0,0,1920,115200...     PL  \n",
       "2  DP=30;I16=27,0,0,0,1200,57600,0,0,1620,97200,0...     PL  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import google.datalab.bigquery as bq\n",
    "query = bq.Query('SELECT * FROM `jbei-cloud.demonstration.variants3` LIMIT 3')\n",
    "output_options = bq.QueryOutput.table(use_cache=False)\n",
    "result = query.execute(output_options=output_options).result()\n",
    "result.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 style=\"color:#9E9E9E\">References</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. McKenna, Aaron, et al. \"The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data.\" Genome research 20.9 (2010): 1297-1303.\n",
    "2. Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., ... & Durbin, R. (2009). The sequence alignment/map format and SAMtools. Bioinformatics, 25(16), 2078-2079."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 style=\"color:#9E9E9E\">Contact</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Want to get in touch? You can [provide feedback](https://goo.gl/forms/2cOmuUrQ3n3CKpim1) regarding this or other documentation,\n",
    "[reach out to us](https://goo.gl/forms/j8FWdNJqABAoJvcW2) regarding collaboration, or [request a new feature or analytical capability](https://goo.gl/forms/dQm3SDcoNZsV7AAd2). We're looking forward to hearing from you!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"padding-top: 30px\">\n",
    "<p style=\"color:#9E9E9E; text-align:center\">This notebook was prepared for <a href=\"https://github.com/cwbeitel/inquiry\">Project Inquiry</a> in support of the research mission of the Joint BioEnergy Institute (JBEI). Learn more at https://www.jbei.org/.</p>\n",
    "<p style=\"color:#9E9E9E; text-align:center\">The Joint BioEnergy Institute is a program of the U.S. Department of Energy Office of Science.</p>\n",
    "<p style=\"color:#9E9E9E; text-align:center\">Â© Regents of the University of California, 2017. Licensed under a BSD-3 <a href=\"https://github.com/cwbeitel/inquiry/blob/master/LICENSE\">license</a>.</p>\n",
    "<img src=\"https://raw.githubusercontent.com/iqtk/iqtk/master/inquiry/docs/assets/logotype_blue_small.png\" style=\"width:100px\"></img>\n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
