{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"padding-bottom:30px\">\n",
    "<a href=\"https://github.com/iqtk/iqtk\"><img src=\"https://raw.githubusercontent.com/iqtk/iqtk/master/inquiry/docs/assets/logotype_blue_small.png\" style=\"width:100px; margin-left:0px\"></img></a>\n",
    "<p style=\"color:#9E9E9E\">\n",
    "<a href=\"https://github.com/cwbeitel/inquiry/tree/master/docs\">Getting Started Guide</a> // <a href=\"https://goo.gl/forms/2cOmuUrQ3n3CKpim1\">Documentation Feedback</a></p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 style=\"color:#9E9E9E\">Metabolomics data preprocessing</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this tutorial we initiate a distributed file conversion of Agilent .d files into .mzML from the DataFlow UI permitting downstream analysis with XCMS3 or alternative tools."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To review, here we'll be running workflows generated by the Inquiry Toolkit directly from the Cloud DataFlow UI so once you've completed the [getting started documentation]() you'll be ready for this tutorial!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 style=\"color:#9E9E9E\">Configuration and Run</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing we need to do is parameterize our analysis. The parameters file we'll need to provide should have the following structure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```json\n",
    "{\n",
    "  \"archives\": [\n",
    "    \"gs://inquiry-test/tests/toolkit-ms/file.d\"\n",
    "   ]\n",
    "}\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Dataflow submission UI screenshot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once our job is submitted we'll be able to check its status using the Cloud DataFow UI as described in the [getting started documentation](). For illustration this workflow should look like the following:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Dataflow job running screenshot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When the workflow is complete you can obtain the path in Google Cloud Storage to the resulting files as described in the [getting started documentation]()."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 style=\"color:#9E9E9E\">Exploring the data</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above workflow results in converted mzML files ready to be used with XCMS3. What follows is a tutorial on obtaining and performing basic analysis of a whole metabolome dataset from BigQuery which was not desposited by the above workflow. First we obtain the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>Factors</th>\n",
       "      <th>_1_2_4_benzenetriol</th>\n",
       "      <th>_1_monostearin</th>\n",
       "      <th>_2_hydroxyvaleric_acid</th>\n",
       "      <th>_3_phosphoglycerate</th>\n",
       "      <th>_5_hydroxynorvaline_NIST</th>\n",
       "      <th>adenosine</th>\n",
       "      <th>adenosine_5_monophosphate</th>\n",
       "      <th>adipic_acid</th>\n",
       "      <th>...</th>\n",
       "      <th>_222169</th>\n",
       "      <th>_222058</th>\n",
       "      <th>_222051</th>\n",
       "      <th>_222047</th>\n",
       "      <th>_222036</th>\n",
       "      <th>_222029</th>\n",
       "      <th>_199553</th>\n",
       "      <th>_199322</th>\n",
       "      <th>_207188</th>\n",
       "      <th>_199232</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LabF_115909</td>\n",
       "      <td>Arabidopsis Genotype:fatb-ko KD; At1g08510 | P...</td>\n",
       "      <td>3566</td>\n",
       "      <td>450</td>\n",
       "      <td>931</td>\n",
       "      <td>2005</td>\n",
       "      <td>2403</td>\n",
       "      <td>407</td>\n",
       "      <td>1650</td>\n",
       "      <td>8148</td>\n",
       "      <td>...</td>\n",
       "      <td>254</td>\n",
       "      <td>358</td>\n",
       "      <td>788</td>\n",
       "      <td>2015</td>\n",
       "      <td>7857</td>\n",
       "      <td>1496</td>\n",
       "      <td>378</td>\n",
       "      <td>1440</td>\n",
       "      <td>916</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LabF_115904</td>\n",
       "      <td>Arabidopsis Genotype:fatb-ko KD; At1g08510 | P...</td>\n",
       "      <td>1874</td>\n",
       "      <td>987</td>\n",
       "      <td>771</td>\n",
       "      <td>2039</td>\n",
       "      <td>2297</td>\n",
       "      <td>538</td>\n",
       "      <td>2153</td>\n",
       "      <td>2197</td>\n",
       "      <td>...</td>\n",
       "      <td>254</td>\n",
       "      <td>602</td>\n",
       "      <td>949</td>\n",
       "      <td>1952</td>\n",
       "      <td>5013</td>\n",
       "      <td>1840</td>\n",
       "      <td>443</td>\n",
       "      <td>1498</td>\n",
       "      <td>980</td>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LabF_115914</td>\n",
       "      <td>Arabidopsis Genotype:fatb-ko KD; At1g08510 | P...</td>\n",
       "      <td>1945</td>\n",
       "      <td>1910</td>\n",
       "      <td>1114</td>\n",
       "      <td>1496</td>\n",
       "      <td>2598</td>\n",
       "      <td>575</td>\n",
       "      <td>2204</td>\n",
       "      <td>2149</td>\n",
       "      <td>...</td>\n",
       "      <td>651</td>\n",
       "      <td>751</td>\n",
       "      <td>721</td>\n",
       "      <td>919</td>\n",
       "      <td>5963</td>\n",
       "      <td>1870</td>\n",
       "      <td>425</td>\n",
       "      <td>1423</td>\n",
       "      <td>1006</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 382 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Samples                                            Factors  \\\n",
       "0  LabF_115909  Arabidopsis Genotype:fatb-ko KD; At1g08510 | P...   \n",
       "1  LabF_115904  Arabidopsis Genotype:fatb-ko KD; At1g08510 | P...   \n",
       "2  LabF_115914  Arabidopsis Genotype:fatb-ko KD; At1g08510 | P...   \n",
       "\n",
       "   _1_2_4_benzenetriol  _1_monostearin  _2_hydroxyvaleric_acid  \\\n",
       "0                 3566             450                     931   \n",
       "1                 1874             987                     771   \n",
       "2                 1945            1910                    1114   \n",
       "\n",
       "   _3_phosphoglycerate  _5_hydroxynorvaline_NIST  adenosine  \\\n",
       "0                 2005                      2403        407   \n",
       "1                 2039                      2297        538   \n",
       "2                 1496                      2598        575   \n",
       "\n",
       "   adenosine_5_monophosphate  adipic_acid   ...     _222169  _222058  _222051  \\\n",
       "0                       1650         8148   ...         254      358      788   \n",
       "1                       2153         2197   ...         254      602      949   \n",
       "2                       2204         2149   ...         651      751      721   \n",
       "\n",
       "   _222047  _222036  _222029  _199553  _199322  _207188  _199232  \n",
       "0     2015     7857     1496      378     1440      916      324  \n",
       "1     1952     5013     1840      443     1498      980      321  \n",
       "2      919     5963     1870      425     1423     1006      228  \n",
       "\n",
       "[3 rows x 382 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import google.datalab.bigquery as bq\n",
    "query = bq.Query('SELECT * FROM `jbei-cloud.metabolomics.feihnexample3` LIMIT 3')\n",
    "output_options = bq.QueryOutput.table(use_cache=False)\n",
    "result = query.execute(output_options=output_options).result()\n",
    "result.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 style=\"color:#9E9E9E\">References</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Trapnell, Cole, et al. \"Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks.\" Nature protocols 7.3 (2012): 562-578.\n",
    "2. Trapnell, Cole, Lior Pachter, and Steven L. Salzberg. \"TopHat: discovering splice junctions with RNA-Seq.\" Bioinformatics 25.9 (2009): 1105-1111.\n",
    "3. Langmead, Ben, and Steven L. Salzberg. \"Fast gapped-read alignment with Bowtie 2.\" Nature methods 9.4 (2012): 357-359."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 style=\"color:#9E9E9E\">Contact</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Want to get in touch? You can [provide feedback](https://goo.gl/forms/2cOmuUrQ3n3CKpim1) regarding this or other documentation,\n",
    "[reach out to us](https://goo.gl/forms/j8FWdNJqABAoJvcW2) regarding collaboration, or [request a new feature or analytical capability](https://goo.gl/forms/dQm3SDcoNZsV7AAd2). We're looking forward to hearing from you!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"padding-top: 30px\">\n",
    "<p style=\"color:#9E9E9E; text-align:center\">This notebook was prepared for <a href=\"https://github.com/cwbeitel/inquiry\">Project Inquiry</a> in support of the research mission of the Joint BioEnergy Institute (JBEI). Learn more at https://www.jbei.org/.</p>\n",
    "<p style=\"color:#9E9E9E; text-align:center\">The Joint BioEnergy Institute is a program of the U.S. Department of Energy Office of Science.</p>\n",
    "<p style=\"color:#9E9E9E; text-align:center\">© Regents of the University of California, 2017. Licensed under a BSD-3 <a href=\"https://github.com/cwbeitel/inquiry/blob/master/LICENSE\">license</a>.</p>\n",
    "<img src=\"https://raw.githubusercontent.com/iqtk/iqtk/master/inquiry/docs/assets/logotype_blue_small.png\" style=\"width:100px\"></img>\n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
