{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"padding-bottom:30px\">\n",
    "<a href=\"https://github.com/iqtk/iqtk\"><img src=\"https://raw.githubusercontent.com/iqtk/iqtk/master/inquiry/docs/assets/logotype_blue_small.png\" style=\"width:100px; margin-left:0px\"></img></a>\n",
    "<p style=\"color:#9E9E9E\">\n",
    "<a href=\"https://github.com/cwbeitel/inquiry/tree/master/docs\">Getting Started Guide</a> // <a href=\"https://goo.gl/forms/2cOmuUrQ3n3CKpim1\">Documentation Feedback</a></p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 style=\"color:#9E9E9E\">Gene expression analysis</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this analysis we perform differential expression analysis with the Cufflinks toolset which includes [cufflinks](https://cole-trapnell-lab.github.io/cufflinks/), [tophat](https://ccb.jhu.edu/software/tophat/), and [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml). You can brush up on gene expression profiling [here](https://en.wikipedia.org/wiki/Gene_expression_profiling).\n",
    "\n",
    "To review, here we'll be running workflows generated by the Inquiry Toolkit directly from the Cloud DataFlow UI so once you've completed the [getting started documentation]() you'll be ready for this tutorial!\n",
    "\n",
    "Our data processing pipeline will deliver data to a BigQuery data table and we'll both explore our data from the BigQuery UI as well as pull a subset into this notebook for subsequent analysis and visualization."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 style=\"color:#9E9E9E\">Configuration and Run</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing we need to do is parameterize our analysis. The parameters file we'll need to provide should have the following structure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```json\n",
    "{\n",
    "  \"ref_fasta\": \"gs://cflow-public/data/genomes/Drosophila_melanogaster/Ensembl/BDGP5.25/Sequence/BowtieIndex/genome.fa\",\n",
    "  \"genes_gtf\": \"gs://cflow-public/data/genomes/Drosophila_melanogaster/Ensembl/BDGP5.25/Annotation/Archives/archive-2015-07-17-14-30-26/Genes/genes.gtf\",\n",
    "  \"cond_a_pairs\": [\n",
    "      [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794483_C1_R1_1_small.fq\",\n",
    "       \"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794483_C1_R1_2_small.fq\"],\n",
    "      [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794484_C1_R2_1_small.fq\",\n",
    "       \"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794484_C1_R2_2_small.fq\"],\n",
    "      [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794485_C1_R3_1_small.fq\",\n",
    "       \"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794485_C1_R3_2_small.fq\"]\n",
    "      ],\n",
    "  \"cond_b_pairs\": [\n",
    "       [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794486_C2_R1_1_small.fq\",\n",
    "        \"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794486_C2_R1_2_small.fq\"],\n",
    "       [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794487_C2_R2_1_small.fq\",\n",
    "        \"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794487_C2_R2_2_small.fq\"],\n",
    "       [\"gs://cflow-public/data/rnaseq/downsampled_reads/GSM794488_C2_R3_1_small.fq\",\n",
    "        \"gs://cflow-public/data/rnaseq/downsampled_readsGSM794488_C2_R3_2_small.fq\"]\n",
    "       ]\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The job can be initiated from the DataFlow custom template job submission UI which should look like the following (TODO):\n",
    "![](https://camo.githubusercontent.com/6557bc51dd5b7d8b55e636ae94298fec7adfee11/68747470733a2f2f636c6f75642e676f6f676c652e636f6d2f64617461666c6f772f696d616765732f776f7264636f756e745f74656d706c6174655f657865637574652e706e67)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once our job is submitted we'll be able to check its status using the Cloud DataFow UI as described in the [getting started documentation](). For illustration this workflow should look like the following:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![](../assets/wf-sshot-transcriptomics.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When the workflow is complete you can obtain the path in Google Cloud Storage to the resulting files as described in the [getting started documentation]()."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2 style=\"color:#9E9E9E\">Exploring the data</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we'll obtain some data resulting from the above workflow and play around a bit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>geneid</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample1</th>\n",
       "      <th>sample2</th>\n",
       "      <th>status</th>\n",
       "      <th>expression1</th>\n",
       "      <th>expression2</th>\n",
       "      <th>lnFoldChange</th>\n",
       "      <th>testStatistic</th>\n",
       "      <th>pValue</th>\n",
       "      <th>qValue</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XLOC_001687</td>\n",
       "      <td>XLOC_001687</td>\n",
       "      <td>l(2)05714</td>\n",
       "      <td>2L:4969003-4971453</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XLOC_014024</td>\n",
       "      <td>XLOC_014024</td>\n",
       "      <td>CG14408</td>\n",
       "      <td>X:14721090-14726090</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XLOC_000602</td>\n",
       "      <td>XLOC_000602</td>\n",
       "      <td>gcm2</td>\n",
       "      <td>2L:9608478-9612710</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id       geneid       gene                locus sample1 sample2  \\\n",
       "0  XLOC_001687  XLOC_001687  l(2)05714   2L:4969003-4971453      C1      C2   \n",
       "1  XLOC_014024  XLOC_014024    CG14408  X:14721090-14726090      C1      C2   \n",
       "2  XLOC_000602  XLOC_000602       gcm2   2L:9608478-9612710      C1      C2   \n",
       "\n",
       "   status expression1 expression2 lnFoldChange testStatistic  pValue  qValue  \\\n",
       "0  NOTEST        None        None         None          None     1.0     1.0   \n",
       "1  NOTEST        None        None         None          None     1.0     1.0   \n",
       "2  NOTEST        None        None         None          None     1.0     1.0   \n",
       "\n",
       "  significant  \n",
       "0        None  \n",
       "1        None  \n",
       "2        None  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import google.datalab.bigquery as bq\n",
    "query = bq.Query('SELECT * FROM `jbei-cloud.somedataset.sometable2` LIMIT 3')\n",
    "output_options = bq.QueryOutput.table(use_cache=False)\n",
    "result = query.execute(output_options=output_options).result()\n",
    "result.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 style=\"color:#9E9E9E\">References</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Trapnell, Cole, et al. \"Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks.\" Nature protocols 7.3 (2012): 562-578.\n",
    "2. Trapnell, Cole, Lior Pachter, and Steven L. Salzberg. \"TopHat: discovering splice junctions with RNA-Seq.\" Bioinformatics 25.9 (2009): 1105-1111.\n",
    "3. Langmead, Ben, and Steven L. Salzberg. \"Fast gapped-read alignment with Bowtie 2.\" Nature methods 9.4 (2012): 357-359."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 style=\"color:#9E9E9E\">Contact</h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Want to get in touch? You can [provide feedback](https://goo.gl/forms/2cOmuUrQ3n3CKpim1) regarding this or other documentation,\n",
    "[reach out to us](https://goo.gl/forms/j8FWdNJqABAoJvcW2) regarding collaboration, or [request a new feature or analytical capability](https://goo.gl/forms/dQm3SDcoNZsV7AAd2). We're looking forward to hearing from you!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"padding-top: 30px\">\n",
    "<p style=\"color:#9E9E9E; text-align:center\">This notebook was prepared for <a href=\"https://github.com/cwbeitel/inquiry\">Project Inquiry</a> in support of the research mission of the Joint BioEnergy Institute (JBEI). Learn more at https://www.jbei.org/.</p>\n",
    "<p style=\"color:#9E9E9E; text-align:center\">The Joint BioEnergy Institute is a program of the U.S. Department of Energy Office of Science.</p>\n",
    "<p style=\"color:#9E9E9E; text-align:center\">© Regents of the University of California, 2017. Licensed under a BSD-3 <a href=\"https://github.com/cwbeitel/inquiry/blob/master/LICENSE\">license</a>.</p>\n",
    "<img src=\"https://raw.githubusercontent.com/iqtk/iqtk/master/inquiry/docs/assets/logotype_blue_small.png\" style=\"width:100px\"></img>\n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
